### PR TITLE
Rework access checker and access restrictions

### PR DESCRIPTION
**Pull request by IvanIvanoff**
## Changes

- Fix issue when `realtime_days_cut_off` is `0` to properly put `null` in `restrictedTo`.
- Fix function args inconsistencies (use `plan, product, metric` order in most places)
- Add caching to get_access_restrictions

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have tried to find clearer solution before commenting hard-to-understand parts of code
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/sanbase_web/graphql/middlewares/access_control.ex [hunk 1]
```diff
@@ -189,10 +187,10 @@ defmodule SanbaseWeb.Graphql.Middlewares.AccessControl do
     product_id = subscription.plan.product_id || context.product_id
 
     historical_data_in_days =
-      Subscription.historical_data_in_days(subscription, query, product_id)
+      Subscription.historical_data_in_days(subscription, product_id, query)
```
Comment (tspenov) [inline lib/sanbase_web/graphql/middlewares/access_control.ex:190]: We can use `AccessChecker.historical_data_in_days(subscription.plan, product_id, query)` and remove the one from `Subscription`. Same for `realtime_data_cut_off_in_days`

File: lib/sanbase_web/graphql/resolvers/access_control_resolver.ex [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
-  alias Sanbase.Billing.Subscription
+  alias Sanbase.Billing.Product
+  alias SanbaseWeb.Graphql.Cache
 
-  def get_access_restrictions(_root, _args, %{
-        context: %{product_id: product_id} = context
-      }) do
-    subscription = context[:auth][:subscription] || Subscription.free_subscription()
+  def get_access_restrictions(_root, args, %{context: context}) do
+    plan = Map.get(args, :plan) || context[:auth][:plan] || :free
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/access_control_resolver.ex:6]: I don't think we pass any args to `get_access_restrictions` so `Map.get(args, :plan)` won't work. But I think we do have it always here `context[:auth][:plan]` :?

File: lib/sanbase_web/graphql/resolvers/access_control_resolver.ex [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
 
-    restrictions = Sanbase.Billing.Plan.Restrictions.get_all(subscription, product_id)
+    product_id =
+      Product.id_by_code(Map.get(args, :product)) || context[:product_id] || Product.product_api()
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/access_control_resolver.ex:9]: Same here.

File: lib/sanbase_web/graphql/resolvers/access_control_resolver.ex (hunk 2 above)
Comment (tspenov) [reply lib/sanbase_web/graphql/resolvers/access_control_resolver.ex:6]: Ok, I didn't bother to scroll below. I saw that we provide plan and product as optional args

File: lib/sanbase_web/graphql/middlewares/access_control.ex (hunk 1 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/middlewares/access_control.ex]: Done

File: lib/sanbase_web/graphql/middlewares/access_control.ex [hunk 4]
```diff
@@ -41,7 +47,7 @@ defmodule SanbaseWeb.Graphql.Middlewares.AccessControl do
   end
 
   # The name of the query/mutation can be passed in snake case or camel case.
-  # Here we transform the name to an atom in snake case for consistency
+  # Here we transform the name to an altom in snake case for consistency
```
Comment (tspenov) [inline lib/sanbase_web/graphql/middlewares/access_control.ex:50]:
```suggestion
  # Here we transform the name to an atom in snake case for consistency
```

**Review by tspenov — APPROVED** [review]

**Pull request merged**